EnoDatabase/EnoDatabase.cs

```diff
@@ -687,14 +687,82 @@ public async Task CalculateRoundTeamServiceStates(IServiceProvider serviceProvid
             var teams = await _context.Teams.AsNoTracking().ToArrayAsync();
             var services = await _context.Services.AsNoTracking().ToArrayAsync();
 
-            var currentRoundWorstResults = await _context.CheckerTasks
+            //Logger.LogError("Before Statement");
+            var currentRoundWorstResults = new Dictionary<(long ServiceId, long TeamId), CheckerTask?>();
+            var sw = new Stopwatch();
+            sw.Start();
+            foreach (var t in teams)
+                foreach (var s in services)
+            {
+                    currentRoundWorstResults[(s.Id, t.Id)] =
+                        await _context.CheckerTasks
+                            .TagWith("CalculateRoundTeamServiceStates:currentRoundTasks")
+                            .Where(ct => ct.CurrentRoundId == roundId)
+                            .Where(ct => ct.RelatedRoundId == roundId)
+                            .Where(ct => ct.TeamId == t.Id)
+                            .Where(ct => ct.ServiceId == s.Id)
+                            .Where(ct => ct.CheckerResult == CheckerResult.INTERNAL_ERROR)
+                            .OrderBy(ct => ct.StartTime)

```

Comment:
What's postgres' take on order bys and composite indexes?

Reply:
You mean if the start time field needs to be indexed? I am not sure, the index definitely needs some more love.

Reply:
[Your index columns must be in the correct order if you want it to be effective, at least if you want to stay with btree](https://www.postgresql.org/docs/8.3/indexes-multicolumn.html). Indexes in general work [with order by](https://www.postgresql.org/docs/10/indexes-ordering.html), but the docs don't explicitly talk about composite indexes there.

You can mimic the group-by-get-first query by using `SELECT DISCTINCT ON` + `ORDER BY` [according to so](https://stackoverflow.com/a/7630564/1569755), but in [pg 8.3 it was slow for huge tables](https://explainextended.com/2009/05/03/postgresql-optimizing-distinct/). However that was more than 10 years ago, it might be worth investigating whether that is still the case.

While benchmarking please remember to use huge tables, otherwise you won't see the effects of your indexes.